### ISSUE TITLE
Alluxio loads ufs jars multiple times during journal formatting

### ISSUE THREAD

**Issue by ChernikovP** [Alluxio/alluxio#12113]
**Alluxio Version:**
I'm working of `branch-2.3`

**Describe the bug**
Formatting journal results in `Native Library /tmp/mapr-root-libMapRClient.unknown.so already loaded in another classloader` error. For full log, please, see https://gist.github.com/ChernikovP/687d9bdcbec2b0f06fe318489ddb45ce.

What's happening is that for every single service: `BlockMaster`, `FileSystemMaster`, `MetaMaster`, `MetricsMaster` and `TableMaster` execution of https://github.com/Alluxio/alluxio/blob/680b2fd4ff13d3349572518f15e1f9f5b2fffa93/core/server/common/src/main/java/alluxio/cli/Format.java#L120 results in loading all possible ufs factories, filtering by path prefix the necessary one and trying to process journal entry

**To Reproduce**
- Add `mapr` repository to `pom.xml`:

```
<repository>
  <id>mapr-releases</id>
  <url>https://repository.mapr.com/maven/</url>
  <snapshots>
    <enabled>false</enabled>
  </snapshots>
  <releases>
    <enabled>true</enabled>
  </releases>
</repository>
```

- add follwoing module to `ufsModules` in `dev/scripts/src/alluxio.org/build-distribution/cmd/common.go`:

```
{"hadoop-mapr", true, "-pl underfs/hdfs -Pufs-hadoop-2 -Dufs.hadoop.version=2.7.0-mapr-1808 -PhdfsActiveSync"}
```

- build binaries by running `./dev/scripts/generate-tarballs single -ufs-modules=ufs-hadoop-mapr`

- conf file:

```
# Underfs setup
alluxio.underfs.hdfs.configuration=/opt/mapr/hadoop/hadoop-2.7.0/etc/hadoop/core-site.xml:/opt/mapr/hadoop/hadoop-2.7.0/etc/hadoop/hdfs-site.xml
alluxio.underfs.hdfs.prefixes=maprfs:/
alluxio.underfs.hdfs.impl=com.mapr.fs.MapRFileSystem
# Master setup
alluxio.master.backup.directory=maprfs:///storage/scratch/alluxio/backup
alluxio.master.journal.folder=maprfs:///storage/scratch/alluxio/journal
alluxio.master.journal.type=UFS
alluxio.master.mount.table.root.ufs=maprfs:///
```

- run `./bin/alluxio format` on Alluxio master

**Expected behavior**
Successful  journal formatting

**Urgency**
Describe the impact and urgency of the bug.

**Additional context**
I tried s3 as ufs and get this eventually:

```
2020-08-18 08:42:05,559 INFO  ExtensionFactoryRegistry - Loading core jars from /opt/alluxio-2.3.1-SNAPSHOT-noHelm/lib
2020-08-18 08:42:05,760 DEBUG ExtensionFactoryRegistry - Factory implementation alluxio.underfs.s3a.S3AUnderFileSystemFactory@9f70c54 is eligible for path s3://chernikovp-alluxio/journal/BlockMaster
2020-08-18 08:42:07,212 INFO  ExtensionFactoryRegistry - Loading core jars from /opt/alluxio-2.3.1-SNAPSHOT-noHelm/lib
2020-08-18 08:42:07,243 DEBUG ExtensionFactoryRegistry - Factory implementation alluxio.underfs.s3a.S3AUnderFileSystemFactory@7d3e8655 is eligible for path s3://chernikovp-alluxio/journal/FileSystemMaster
2020-08-18 08:42:07,587 INFO  ExtensionFactoryRegistry - Loading core jars from /opt/alluxio-2.3.1-SNAPSHOT-noHelm/lib
2020-08-18 08:42:07,694 DEBUG ExtensionFactoryRegistry - Factory implementation alluxio.underfs.s3a.S3AUnderFileSystemFactory@346939bf is eligible for path s3://chernikovp-alluxio/journal/MetaMaster
2020-08-18 08:42:08,058 INFO  ExtensionFactoryRegistry - Loading core jars from /opt/alluxio-2.3.1-SNAPSHOT-noHelm/lib
2020-08-18 08:42:08,123 DEBUG ExtensionFactoryRegistry - Factory implementation alluxio.underfs.s3a.S3AUnderFileSystemFactory@62ddd21b is eligible for path s3://chernikovp-alluxio/journal/MetricsMaster
2020-08-18 08:42:08,432 INFO  ExtensionFactoryRegistry - Loading core jars from /opt/alluxio-2.3.1-SNAPSHOT-noHelm/lib
2020-08-18 08:42:08,577 DEBUG ExtensionFactoryRegistry - Factory implementation alluxio.underfs.s3a.S3AUnderFileSystemFactory@474c9131 is eligible for path s3://chernikovp-alluxio/journal/TableMaster
```
Basically, jars are loading multiple times regardless of underlying fs, but it seems like it leads to errors only in my case...

**Comment by apc999** [comment]
@ChernikovP  thanks for reporting this issue, as you analyzed here, it is because each logic master (FileSystemMaster, BlockMaster and etc, all belonging to one AlluxioMasterProcess) creates a UFS journal location here https://github.com/Alluxio/alluxio/blob/branch-2.3/core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java#L149
using `UnderFileSystem.Factory.create` to initialize a new `UnderFileSystem` Instance.

## Workaround

@ChernikovP Can you set `alluxio.master.mount.table.root.ufs` to your local disk rather than maprfs, but mount your maprfs into Alluxio namespace. In this way, journals will be write to master local storage without causing the conflict loading maprfs jar, whereas you can still access your data in maprfs in nested mount point.

## Clean Fix

Technically, we should share the `UnderFileSystem` not only for the purpose of writing journal among different masters, but also for other part in master. In `alluxio.master.AbstractMaster` we already have a handle to `UfsManager`. So when creating journal `mJournal = masterContext.getJournalSystem().createJournal(this);` in `AbstractMaster` we should pass UfsManager.

Here is a related code change I made before to consolidate different `UnderFileSystem` instances https://github.com/Alluxio/alluxio/pull/9808/files

**Comment by maobaolong** [comment]
@apc999 Through the issue description, your comments and following debug snapshot, I found the key point of this issue is `alluxio formatMaster` shouldn't create multiply instance. 

![image](https://user-images.githubusercontent.com/17329931/96846439-7ec4d700-1484-11eb-8944-796fd3335660.png)

![image](https://user-images.githubusercontent.com/17329931/96846580-afa50c00-1484-11eb-9b82-546a32c0fcc5.png)

![image](https://user-images.githubusercontent.com/17329931/96846694-d105f800-1484-11eb-8e1c-18139bb3ec31.png)

So, for this issue, we can focus on make `alluxio formatMaster` successful and only `load jars` once.
